### PR TITLE
catbuf nem handling

### PR DESCRIPTION
**Pull request by gimre-xymcity**
Each comment is a separate issues, corresponding schemas https://github.com/symbol/catbuffer-schemas/commit/dd508e602caa015a57173056a36af8a590289a26

### CODE REVIEW

File: generator/main.py [hunk 1]
```diff
@@ -35,6 +34,7 @@ def __init__(self, yaml_descriptor, field_name, type_instance, printer):
         self.type_instance = type_instance
         self.printer = printer
         self.bound_field = None
+        self.size_fields = []
```
Comment (Jaguar0625) [inline generator/main.py:37]: why list ?

File: generator/main.py [hunk 2]
```diff
@@ -91,16 +94,46 @@ def create_printer(type_instance, name):
...
-        fix_item(factory_type.yaml_descriptor, 'discriminator')
+def process_concrete_struct(type_map, struct_type, abstract_impl_map):
+    factory_type = type_map[struct_type.yaml_descriptor['factory_type']]
+
+    discriminators = factory_type.yaml_descriptor.get('discriminator', [])
+    for i, discriminator in enumerate(discriminators):
+        discriminators[i] = fix_name(discriminator)
+
+    initializers = factory_type.yaml_descriptor.get('initializers', None)
+    for init in initializers:
```
Comment (Jaguar0625) [inline generator/main.py:105]: probably don't need get, since you can enumerate None

File: generator/printers.py [hunk 3]
```diff
@@ -190,6 +190,12 @@ def get_size(self):
         return f'self.{self.name}.size()'
 
     def load(self, buffer_name='buffer_'):
+        if self.descriptor.is_struct and self.descriptor.is_abstract:
+            # HACK: factories use this printers as well, ignore them
+            if 'parent' != self.name:
```
Comment (Jaguar0625) [inline generator/printers.py:195]: when is `self.name == 'parent'` ?

File: generator/StructTypeFormatter.py [hunk 4]
```diff
@@ -130,26 +130,45 @@ def generate_condition(self, field, prefix_field=False):
...
         }
         condition_operator = condition_to_operator_map[field.yaml_descriptor['condition_operation']]
-        yoda_value = f'{condition_field.get_type().typename}.{field.yaml_descriptor["condition_value"]}'
+
+        value = f'{field.yaml_descriptor["condition_value"]}'
+        yoda_value = value if condition_field.get_type().is_int else f'{condition_field.get_type().typename}.{value}'
         field_prefix = 'self.' if prefix_field else ''
+
+        # HACK: instead of handling dumb magic value in namespace parent_name, generate slightly simpler condition
+        if prefix_field and field.yaml_descriptor.get('disposition', '').startswith('array'):
```
Comment (Jaguar0625) [inline generator/StructTypeFormatter.py:139]: unsure if you can use `is_array`

File: generator/type_objects.py [hunk 5]
```diff
@@ -6,15 +6,16 @@ class BaseObject:
...
     CAMEL_CASE_PATTERN = re.compile(r'(?<!^)(?=[A-Z])')
 
-    def __init__(self, base_typename, yaml_descritor):
+    def __init__(self, base_typename, yaml_descriptor):
         self.base_typename = base_typename
         self.is_struct = 'struct' == base_typename
         self.is_array = 'array' == base_typename
         self.is_int = 'int' == base_typename
         self.is_enum = 'enum' == base_typename
+        self.is_sizeof = 'disposition' in yaml_descriptor and 'sizeof' == yaml_descriptor['disposition'] and yaml_descriptor['value']
```
Comment (Jaguar0625) [inline generator/type_objects.py:15]: can use get

Comment (Jaguar0625) [reply generator/type_objects.py:15]: why do you need `yaml_descriptor['value']` check ?

File: generator/main.py (hunk 1 above)
Comment (gimre-xymcity) [reply generator/main.py:37]: this is complementary to `bound_field`

`bound_field` is for size fields in which case `bound_field` points to struct/array

right now I'm using size_fields only for `sizeof` (technically there's nothing that would stop from
having two sizeof fields 😬, I'm only using [0] element anyway)

File: generator/printers.py (hunk 3 above)
Comment (gimre-xymcity) [reply generator/printers.py:195]: as comment says, this happens in FactoryFormatter as it is reusing BuiltinPrinter:

`self.printer = BuiltinPrinter(abstract_type_instance, 'parent')`

File: generator/StructTypeFormatter.py (hunk 4 above)
Comment (gimre-xymcity) [reply generator/StructTypeFormatter.py:139]: is_array is a bit too wide (it covers built in types as well),
and in this case it ends up changing some of _proper_ conditions in symbol, sample

```py
        if self.linked_public_key:
            size += self.linked_public_key.size()
        if self.node_public_key:
            size += self.node_public_key.size()
        if self.vrf_public_key:
            size += self.vrf_public_key.size()
```

vs proper:
```py
        if AccountKeyTypeFlags.LINKED in self.supplemental_public_keys_mask:
            size += self.linked_public_key.size()
        if AccountKeyTypeFlags.NODE in self.supplemental_public_keys_mask:
            size += self.node_public_key.size()
        if AccountKeyTypeFlags.VRF in self.supplemental_public_keys_mask:
            size += self.vrf_public_key.size()
```

File: generator/type_objects.py (hunk 5 above)
Comment (gimre-xymcity) [reply generator/type_objects.py:15]: > why do you need `yaml_descriptor['value']` check ?

seems to be bug, maybe it was missing at some point 🤔

Comment (gimre-xymcity) [reply generator/type_objects.py:15]: heh, it's not a check 😅  (it's setting `is_sizeof` to yaml["value"] - changed to normal if statement)
(and renamed)

**Review by Jaguar0625 — APPROVED** [review]

**Comment by gimre-xymcity** [comment]
merged into feature branch in 9da77dd74138813f629a557b7226df618d1ba39f

**Pull request closed without merging**